File: countColors/as-dataview/countup.ts
export function countColors(uint8Data: Uint8Array): Uint32Array {
  const colorCount = new Uint32Array(16777216);
  const dataView = new DataView(uint8Data.buffer);
  for (let i = 0; i < uint8Data.length; i += 4) {
    const rgba = dataView.getUint32(i, true);
    const rgb = rgba & 0xFFFFFF;
    colorCount[rgb]++;
  }
  return colorCount;
}

File: countColors/go-simple/countup.go
package main

import (
	"unsafe"
)

//export countColors
func countColors(uint8DataPtr *uint8, length int) *uint32 {
	colorCount := make([]uint32, 16777216)
	uint32Data := unsafe.Slice((*uint32)(unsafe.Pointer(uint8DataPtr)), length/4)
	for _, rgba := range uint32Data {
		rgb := rgba & 0xFFFFFF
		colorCount[rgb]++
	}
	return &colorCount[0]
}

func main() {
}
